java/ql/src/Telemetry/AutomodelEndpointCharacteristics.qll

```diff
@@ -0,0 +1,325 @@
+    type instanceof AutomodelEndpointTypes::SqlSinkType
+    or
+    label = "open-url" and
+    humanReadableLabel = "open url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "jdbc-url" and
+    humanReadableLabel = "jdbc url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "command-injection" and
+    humanReadableLabel = "command injection" and
+    type instanceof AutomodelEndpointTypes::CommandInjectionSinkType
+  }
+
+  predicate isSink(Endpoint e, string label) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      string input
+    |
+      sinkSpec(e, package, type, subtypes, name, signature, ext, input) and
+      ExternalFlow::sinkModel(package, type, subtypes, name, [signature, ""], ext, input, label, _)
+    )
+  }
+
+  predicate isNeutral(Endpoint e) {
+    exists(string package, string type, string name, string signature |
+      sinkSpec(e, package, type, _, name, signature, _, _) and
+      ExternalFlow::neutralModel(package, type, name, [signature, ""], _)
+    )
+  }
+
+  additional predicate sinkSpec(
+    Endpoint e, string package, string type, boolean subtypes, string name, string signature,
+    string ext, string input
+  ) {
+    package = e.getEnclosingCallable().getDeclaringType().getPackage().toString() and
+    type = e.getEnclosingCallable().getDeclaringType().getName() and
+    subtypes = false and
+    name = e.getEnclosingCallable().getName() and
+    signature = ExternalFlow::paramsString(e.getEnclosingCallable()) and
+    ext = "" and
+    exists(int paramIdx | e.isParameterOf(_, paramIdx) | input = "Argument[" + paramIdx + "]")
+  }
+
+  predicate hasMetadata(Endpoint e, string metadata) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      int input, string provenance, boolean isPublic, boolean isFinal, boolean isStatic,
+      string calleeJavaDoc
+    |
+      hasMetadata(e, package, type, name, signature, input, isFinal, isStatic, isPublic,
+        calleeJavaDoc) and
+      (if isFinal = true or isStatic = true then subtypes = false else subtypes = true) and
+      ext = "" and // see https://github.slack.com/archives/CP9127VUK/p1673979477496069
+      provenance = "ai-generated" and
+      metadata =
+        "{" //
+          + "'Package': '" + package //
+          + "', 'Type': '" + type //
+          + "', 'Subtypes': " + subtypes //
+          + ", 'Name': '" + name //
+          + ", 'ParamName': '" + e.toString() //
+          + "', 'Signature': '" + signature //
+          + "', 'Ext': '" + ext //
+          + "', 'Argument index': " + input //
+          + ", 'Provenance': '" + provenance //
+          + "', 'Is public': " + isPublic //
+          + "', 'Callee JavaDoc': '" + calleeJavaDoc.replaceAll("'", "\"") //
+          + "'}" // TODO: Why are the curly braces added twice?
```

Comment:
I see the `Is passed to external API` has been dropped, presumably because it would be always `False`, but could we keep it so that we can easily implement the candidates filtering proposal in https://github.com/github/CodeML/issues/42#issuecomment-1529735484?

Reply:
I'd have to add a bogus value to the code that makes no sense in the context of its behaviour — is there a reason you can't use a default value when the value is absent in the python code?

Reply:
Generally I think we should have the examples/candidates extraction queries follow a standard minimum output format. For instance we need to make sure we always have the information necessary to compute the MaD declaration and it's better to be explicit about the values of these fields so that the output of the query has all the information.

Reply:
see also https://github.com/github/codeql/pull/12830#issuecomment-1531769154

Reply:
> Generally I think we should have the examples/candidates extraction queries follow a standard minimum output format.

I agree with that, but I'd like to not add it to this PR to keep its scope from growing too much.

Reply:
Thanks! Happy to defer to you and @adityasharad about the best way to implement this change. Here is the issue I created to track this: https://github.com/github/CodeML/issues/102